test/e2e/scheduler_predicates.go

```diff
@@ -371,4 +371,74 @@ var _ = Describe("SchedulerPredicates", func() {
 		verifyResult(c, podName, ns, currentlyDeadPods)
 		cleanupPods(c, ns)
 	})
+
+	It("validates that NodeSelector is respected if matching.", func() {
+		// launch a pod to find a node which can launch a pod. We intentionally do
+		// not just take the node list and choose the first of them. Depending on the
+		// cluster and the scheduler it might be that a "normal" pod cannot be
+		// scheduled onto it.
+		By("Trying to launch a pod without a label to get a node which can launch it.")
+		podName := "without-label"
+		_, err := c.Pods(ns).Create(&api.Pod{
+			TypeMeta: unversioned.TypeMeta{
+				Kind: "Pod",
+			},
+			ObjectMeta: api.ObjectMeta{
+				Name: podName,
+			},
+			Spec: api.PodSpec{
+				Containers: []api.Container{
+					{
+						Name:  podName,
+						Image: "gcr.io/google_containers/pause:go",
+					},
+				},
+			},
+		})
+		expectNoError(err)
+		expectNoError(waitForPodRunningInNamespace(c, podName, ns))
+		pod, err := c.Pods(ns).Get(podName)
+		expectNoError(err)
+
+		nodeName := pod.Spec.NodeName
+		err = c.Pods(ns).Delete(podName, api.NewDeleteOptions(0))
+		expectNoError(err)
+
+		By("Trying to apply a random label on the found node.")
+		k := fmt.Sprintf("e2e.kubernetes.io/%s", string(util.NewUUID()))
+		v := "42"
+		runKubectl("patch", "node", nodeName, "-p", fmt.Sprintf(`{"metadata":{"labels":{"%s":"%s"}}}`, k, v))
```

Comment:
It's just a matter of taste, but I'd rather have it done by client instead of runKubectl call. You can ignore this comment if it's too much hustle.